designs/0005-static-matrices.md

```diff
@@ -0,0 +1,147 @@
+- Stan Issue: [#1805](https://github.com/stan-dev/math/issues/1805)
+
+# Summary
+[summary]: #summary
+
+This proposes a constant matrix type in stan that will allow for significant performance opportunities. In Stan Math this will be represented by `var<Eigen::Matrix<double, R, C>>` or `var<matrix_cl<double>>`. The implementation will be fully backwards compatible with current Stan code. With optimizations from the compiler the conditions for a `matrix` to be a constant matrix can be detected and applied to Stan programs automatically for users. The implementation of this proposal suggests a staged approach where a stan language level feature is delayed until type attributes are allowed in the language and constant matrices have support for the same set of methods that the current `matrix` type has.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, a differentiable matrix in Stan is represented as an Eigen matrix holding a pointer to an underlying array of autodiff objects, each of those holding a pointer to the underlying autodiff object implementation. These `N*M` autodiff elements are expensive but necessary so the elements of a matrix can be assigned to without copying the entire matrix. However, in instances where the matrix is treated as a whole object such that underlying subsets of the matrix are never assigned to, we can store one autodiff object representing the entire matrix. See [here](https://github.com/stan-dev/design-docs/pull/21#issuecomment-625352581) for performance tests on a small subset of gradient evaluations using matrices vs. constant matrices.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+At the Stan Math level a constant matrix is a `var_value<Eigen::Matrix<double, -1, -1>>` with an underlying pointer to a `vari_value<Eigen::Matrix<double, -1, -1>>`\*.  This means that accessors for the value and adjoint of `var_value` objects can access contiguous chunks of each part of the `vari_value`. Any function that accepts a `var_value<T>` will support static matrices.
+
+At the language and compiler level, a `matrix` can be substituted for a constant matrix if the matrix is only constructed once and it's subslices are never assigned to.
+
+```stan
+const matrix[N, M] A = // Construct matrix A...
+A[10, 10] = 5; // illegal!
+A[1:10, ] = 5; // illegal!
+A[, 1:10] = 5; // illegal!
+A[1:10, 1:10] = 5; // illegal!
+```
+
+Any function or operation in the stan language that can accepts a `matrix` as an argument can also accept a constant matrix. Functions which can take in multiple matrices will only return a constant matrix if all of the other matrix inputs as well as the return type are static matrices.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This feature will be added in the following stages.
+
+1. Replace `var` and `vari` with templated types.
```

Comment:
This part needs updated.

Vaguely we have:

1. var has been replaced by var_value<double> in some branch

2. var_value<Eigen> and var_value<Sparse> types exist

3. We agreed in a math meeting to develop this in pieces. What this means is pull requests will merge into develop before everything that will be developed is finished (https://discourse.mc-stan.org/t/monthly-math-development-meeting-06-25-2020-10-am-edt/16012/23).

It might be helpful to lay out what these pieces look like (what currently exists and what is likely to come in the next 3 months).

We'll basically be rewriting all of reverse mode to support this. Hopefully that can all go through adj_jac_apply.

It's also probably worth discussing return types. I'm still a fan of `f(A, B, C) -> D`, D is a `var_value<Eigen>` type if any of A, B, or C are `Eigen<var>` or `var_value<Eigen>` types.

Should also talk about the status of the LTO Windows/g++/clang++/Boost whatever it is

Reply:
> It's also probably worth discussing return types. I'm still a fan of `f(A, B, C) -> D`, `D` is a `var_value<Eigen>` type if any of A, B, or C are `Eigen<var>` or `var_value<Eigen>` types.

Right now I have

> Functions which can take in multiple matrices will only return a constant matrix if all of the other matrix inputs as well as the return type are static matrices.

I'd rather be conservative on this for now aka having constant matrices stricly happen if all input types and the output type are going to be a constant. We can parse through a program to understand whether that's true or not. Once we run more performance tests to see if the copy of static->dynamic is expensive or not we can relax this